### DISCUSSION TITLE
Get all defined Variables in the current context programatically?

### DISCUSSION THREAD

**Discussion by efficienthacks** [dotnet/interactive#3044] [Q&A]
Is there any way to get the Variables programatically (i.e. in a C# cell)?  (I know you see it in the Polyglot Notebook: Variables pane)

I want to run a different code path depending on whether a variable exists in the Notebook.

Thanks

**Comment by jonsequitur** [comment]
Here's an example of the API used to query for which variables are declared in a given kernel:

```csharp
using Microsoft.DotNet.Interactive;
using Microsoft.DotNet.Interactive.Commands;

var result = await Kernel.Root.SendAsync(new RequestValueInfos("csharp"));
result.Events
```

The `RequestValueInfos` command will respond with a `ValueInfosProduced` event that provides the names and a summary formatting of the value for every variable in the specified subkernel. This is the same command used by the Variables pane.

![image](https://github.com/dotnet/interactive/assets/547415/edbc98c3-5b0a-45b7-b024-0485b0f237cf)

To get the value of a specific variable you can use the `RequestValue` command:

```csharp
var result = await Kernel.Root.SendAsync(new RequestValue("x", mimeType: "application/json", targetKernelName: "csharp"));
result.Events
```

![image](https://github.com/dotnet/interactive/assets/547415/15b63f43-c1db-4de6-9e9c-629db7793be3)

It's also worth noting that the `ValueProduced` event has a `Value` property that's not visible here because it's not serialized. When handling events from an in-process subkernel, this property contains a reference to the actual, unserialized value.
